File: engine/src/flutter/lib/web_ui/lib/src/engine/compositor/picture_recorder.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// @dart = 2.10
part of engine;

class CkPictureRecorder implements ui.PictureRecorder {
  ui.Rect? _cullRect;
  SkPictureRecorder? _skRecorder;
  CkCanvas? _recordingCanvas;

  CkCanvas beginRecording(ui.Rect bounds) {
    _cullRect = bounds;
    final SkPictureRecorder recorder = _skRecorder = SkPictureRecorder();
    final SkRect skRect = toSkRect(bounds);
    final SkCanvas skCanvas = recorder.beginRecording(skRect);
    return _recordingCanvas = CkCanvas(skCanvas);
  }

  CkCanvas? get recordingCanvas => _recordingCanvas;

  @override
  ui.Picture endRecording() {
    final SkPictureRecorder? recorder = _skRecorder;

    if (recorder == null) {
      throw StateError('PictureRecorder is not recording');
    }

    final SkPicture skPicture = recorder.finishRecordingAsPicture();
    recorder.delete();
    _skRecorder = null;
    return CkPicture(skPicture, _cullRect);
  }

  @override
  bool get isRecording => _skRecorder != null;
}

File: engine/src/flutter/impeller/aiks/picture_recorder.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "flutter/impeller/aiks/picture_recorder.h"

#include "flutter/impeller/aiks/canvas.h"

namespace impeller {

PictureRecorder::PictureRecorder() : canvas_(std::make_shared<Canvas>()) {}

PictureRecorder::~PictureRecorder() = default;

std::shared_ptr<Canvas> PictureRecorder::GetCanvas() const {
  return canvas_;
}

std::shared_ptr<Picture> PictureRecorder::EndRecordingAsPicture() {
  return canvas_->EndRecordingAsPicture();
}

}  // namespace impeller
